### PR TITLE
Facebook

### PR DESCRIPTION
**Pull request by MichaelPetrinolis**
Facebook module ->Facebook Login integration

### CODE REVIEW

**Comment by sebastienros** [comment]
Do you want to do a demo next Tuesday?

**Comment by MichaelPetrinolis** [comment]
Thank you for the invitation, I would love to do so. 
My only concern is that this will be my first time.

**Comment by Skrypt** [comment]
There's a first time for everything. Better sooner than later 😉

File: src/OrchardCore.Modules/OrchardCore.Facebook/AdminMenu.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        public IStringLocalizer T { get; set; }
+
+        public void BuildNavigation(string name, NavigationBuilder builder)
+        {
+            if (!String.Equals(name, "admin", StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            builder.Add(T["Facebook"], "15", category =>
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Facebook/AdminMenu.cs]: Under `Configuration`

File: src/OrchardCore.Modules/OrchardCore.Facebook/OrchardCore.Facebook.csproj [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Entities.DisplayManagement\OrchardCore.Entities.DisplayManagement.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.ContentManagement.Abstractions\OrchardCore.ContentManagement.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Data.Abstractions\OrchardCore.Data.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.DisplayManagement\OrchardCore.DisplayManagement.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Environment.Navigation\OrchardCore.Environment.Navigation.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
+    <ProjectReference Include="..\OrchardCore.Roles\OrchardCore.Roles.csproj" />
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Facebook/OrchardCore.Facebook.csproj]: Why do you need a reference to Roles and Users. Ideally you should only need their abstractions.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Facebook/OrchardCore.Facebook.csproj]: Same for Settings.

File: src/OrchardCore.Modules/OrchardCore.Facebook/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# OrchardCore.Facebook
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Facebook/README.md]: A new entry is required in the TOC file.

File: src/OrchardCore/OrchardCore.Users.Core/Services/UserService.cs [hunk 4]
```diff
@@ -12,13 +12,13 @@ namespace OrchardCore.Users.Services
     /// <summary>
     /// Implements <see cref="IUserService"/> by using the ASP.NET Core Identity packages.
     /// </summary>
-    public class UserService : IUserService
+    public class UserService : UserClaimsPrincipalFactory<IUser>, IUserService
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Users.Core/Services/UserService.cs]: Why here and not the `UserManager` ?

Comment (MichaelPetrinolis) [reply src/OrchardCore/OrchardCore.Users.Core/Services/UserService.cs]: I wanted to use the default implementation of UserClaimsPrincipalFactory<TUser> and it depends on UserManager<TUser> and IOptions<IdentityOptions> , so I selected UserService which has the same dependencies.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Manifest.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+[assembly: Feature(
+    Id = FacebookConstants.Features.Login,
+    Name = "Facebook Login",
+    Category = "Facebook",
+    Description = "Auhenticates users from facebook.",
+    Dependencies = new[] { FacebookConstants.Features.Core }
+)]
+
+[assembly: Feature(
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Manifest.cs]: You don't need this feature. If you don't flag its services with [OrchardFeature] then they are assigned to the module's default feature.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookCoreService.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+using System.Threading.Tasks;
+using OrchardCore.Facebook.Settings;
+
+namespace OrchardCore.Facebook.Services
+{
+    public interface IFacebookCoreService
+    {
+        Task<FacebookCoreSettings> GetSettingsAsync();
+        Task UpdateSettingsAsync(FacebookCoreSettings settings);
+        Task<ImmutableArray<ValidationResult>> ValidateSettingsAsync(FacebookCoreSettings settings);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Services/IFacebookCoreService.cs]: Just require `IEnumerable` in interfaces. No need for `ImmutableArray`.

File: src/OrchardCore/OrchardCore.Users.Core/Services/UserService.cs [hunk 7]
```diff
@@ -203,5 +203,27 @@ public void ProcessValidationErrors(IEnumerable<IdentityError> errors, User user
                 }
             }
         }
+
+        #region UserClaimsPrincipalFactory<IUser>
+
+        protected override async Task<ClaimsIdentity> GenerateClaimsAsync(IUser user)
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Users.Core/Services/UserService.cs]: More questions on this one, and summoning @PinpointTownes  too

- Could we have it in a separate class instead, dedicated to this interface?
- Should this implementation be part of the facebook module, or the users module? Meaning are these claims for facebook only, or is it something more general?
- Assuming this is not only for facebook, do we have to use these claims identifier or should/can we use OpenId ones like here: https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/UserInfoController.cs#L65
- When we resolve this interface (we already do in MembershipService) should we resolve an enumeration?
- I don't see any usage of this interface, what component is using it right now, and how is it relevant for Facebook?

I think these questions will trigger more refactoring, at least for `IMembershipService` which I think is using it incorrectly, and for the other services which should actually use it instead of `IMembershipService`.

**Comment by agriffard** [comment]
@MichaelPetrinolis Can you please fix the conflicts?

**Comment by MichaelPetrinolis** [comment]
@agriffard done.

**Comment by KANekT** [comment]
When you planed finish FB login and close / merge in beta 3 this issue?

**Comment by MichaelPetrinolis** [comment]
That depends. There are #1786 and #2029 which are related and might change the way third party authentication services are implemented and probably that's the reason that is not yet merged.

**Comment by KANekT** [comment]
oh... #1786 olaned in rc :(

**Comment by MichaelPetrinolis** [comment]
Don't know if this is the reason not merged. I tried to make a guess.

**Comment by sebastienros** [comment]
I was just worried by the dependencies between components, I need to re-focus on this PR to understand why and if they are necessary. Also it is related to the Layout changes from @MatthijsKrempel

**Comment by MatthijsKrempel** [comment]
@sebastienros had a look at the code, I don't see any overlap,.

**Comment by sebastienros** [comment]
@MichaelPetrinolis I updated your PR, do you want to review my changes? If you are ok with these then I will merge.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs [hunk 8]
```diff
@@ -58,39 +54,28 @@ public override async Task<IDisplayResult> EditAsync(FacebookCoreSettings settin
 
             return Initialize<FacebookCoreSettingsViewModel>("FacebookCoreSettings_Edit", model =>
             {
+                var protector = _dataProtectionProvider.CreateProtector(FacebookConstants.Features.Core);
+
                 model.AppId = settings.AppId;
-                model.AppSecret = settings.AppSecret;
+                model.AppSecret = protector.Unprotect(settings.AppSecret);
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs]: `try/catch` like you did for SMTP and OpenID settings, to avoid decryption exceptions?

File: src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs [hunk 9]
```diff
@@ -58,39 +54,28 @@ public override async Task<IDisplayResult> EditAsync(FacebookCoreSettings settin
...
 
             }).Location("Content:0").OnGroup(FacebookConstants.Features.Core);
         }
 
         public override async Task<IDisplayResult> UpdateAsync(FacebookCoreSettings settings, BuildEditorContext context)
         {
-            var user = _httpContextAccessor.HttpContext?.User;
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs]: Should the other display driver be updated to perform the authorization check at the same stage?

File: src/OrchardCore.Modules/OrchardCore.Facebook/Services/FacebookCoreService.cs [hunk 10]
```diff
@@ -46,32 +46,32 @@ public async Task UpdateSettingsAsync(FacebookCoreSettings settings)
             await _siteService.UpdateSiteSettingsAsync(container);
         }
 
-        public Task<ImmutableArray<ValidationResult>> ValidateSettingsAsync(FacebookCoreSettings settings)
+        public Task<IEnumerable<ValidationResult>> ValidateSettingsAsync(FacebookCoreSettings settings)
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Services/FacebookCoreService.cs]: Out of curiosity, why these changes?

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Services/FacebookCoreService.cs]: I don't see the point of requiring an immutable array on the interface. I remember you had explained why you use them, please remind me.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs (hunk 8 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs]: It's not necessary in drivers, as they are themselves skipped on exceptions and the exception is logged. It is necessary during the configuration phase though, so I will assume that's what you meant.

File: src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs (hunk 9 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs]: I looked for other patterns, but could only find the same one. I assume you are talking about having the security check inside the shape lambda.

**Comment by MichaelPetrinolis** [comment]
I saw two issues, can I commit  in this PR or should I create a new one? 
1.The update for core settings does not work(it checks the settings validity without assigning the values from model)
2. When the appsecret is not set, the edit display driver fails due to unprotecting null string ….

File: src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs (hunk 9 above)
Comment (PinpointTownes) [reply src/OrchardCore.Modules/OrchardCore.Facebook/Drivers/FacebookCoreSettingsDisplayDriver.cs]: No, I meant why moving the authorization check inside the `if` block? It's not consistent with what exists elsewhere (but maybe it's the right thing to do, I dunno): 

https://github.com/OrchardCMS/OrchardCore/blob/80670ba7c474f0c0bae30b22bfbdd3f9bc82e95e/src/OrchardCore.Modules/OrchardCore.Email/Drivers/SmtpSettingsDisplayDriver.cs#L72-L82

File: src/OrchardCore.Modules/OrchardCore.Facebook/Settings/FacebookLoginSettings.cs [hunk 11]
```diff
@@ -0,0 +1,7 @@
+namespace OrchardCore.Facebook.Settings
+{
+    public class FacebookLoginSettings
+    {
+        public string CallbackPath { get; set; }
```
Comment (PinpointTownes) [inline src/OrchardCore.Modules/OrchardCore.Facebook/Settings/FacebookLoginSettings.cs]: Using `PathString` like we now do in the OpenID module could have been a good idea :smile:

**Comment by MichaelPetrinolis** [comment]
Thanx , i will double check it in dev branch and create a new PR if necessary
